src/gwt/src/org/rstudio/core/client/widget/FileOrUrlChooserTextBox.java

```diff
@@ -16,45 +16,79 @@
 package org.rstudio.core.client.widget;
 
 import org.rstudio.core.client.files.FileSystemItem;
+import org.rstudio.studio.client.RStudioGinjector;
 
 import com.google.gwt.event.dom.client.ChangeEvent;
 import com.google.gwt.event.dom.client.ChangeHandler;
+import com.google.gwt.event.dom.client.ClickEvent;
 import com.google.gwt.event.dom.client.ClickHandler;
-import com.google.gwt.event.dom.client.KeyPressEvent;
-import com.google.gwt.event.dom.client.KeyPressHandler;
 import com.google.gwt.event.logical.shared.ValueChangeEvent;
 import com.google.gwt.event.logical.shared.ValueChangeHandler;
 import com.google.gwt.event.shared.HandlerRegistration;
+import com.google.gwt.user.client.Timer;
 import com.google.gwt.user.client.ui.Focusable;
 
-public class FileOrUrlChooserTextBox extends FileChooserTextBox
+public class FileOrUrlChooserTextBox extends TextBoxWithButton
 {
-
-   public FileOrUrlChooserTextBox(String label, Focusable focusAfter)
-   {
-      super(label, focusAfter);
-      
-      setReadOnly(false);
-   }
+   private static String browseModeCaption_ = "Browse...";
+   private static String updateModeCaption_ = "Update";
+   private Boolean updateMode_ = false;
```

Comment:
You probably want a lower-case `boolean` here (`Boolean` is a boxed type and is nullable)